lib/connections/limiter.go

```diff
@@ -280,33 +271,45 @@ func (w *limitedWriter) Write(buf []byte) (int, error) {
 	return written, nil
 }
 
+// waiterHolder is the common functionality around having and evaluating a
+// waiter, valid for both writers and readers
+type waiterHolder struct {
+	waiter    waiter
+	limitsLAN *atomicBool
+	isLAN     bool
+}
+
+// unlimited returns true if the waiter is not limiting the rate
+func (w waiterHolder) unlimited() bool {
+	if w.isLAN && !w.limitsLAN.get() {
+		return true
+	}
+	return w.waiter.Limit() == rate.Inf
+}
+
 // take is a utility function to consume tokens from a overall rate.Limiter and deviceLimiter.
 // No call to WaitN can be larger than the limiter burst size so we split it up into
 // several calls when necessary.
-func take(waiter waiter, tokens int) {
+func (w waiterHolder) take(tokens int) {
 	if tokens < limiterBurstSize {
 		// This is the by far more common case so we get it out of the way
 		// early.
-		waiter.WaitN(context.TODO(), tokens)
+		w.waiter.WaitN(context.TODO(), tokens)
 		return
 	}
 
 	for tokens > 0 {
 		// Consume limiterBurstSize tokens at a time until we're done.
 		if tokens > limiterBurstSize {
```

Comment:
This stuff will now never happen for the writer, but it's *possible* for a read() to return more data than the limiter burst so we still need this looping function

Reply:
Worth changing the comment?